### PR TITLE
Update backend.php

### LINKED ISSUE #350: Not possible to create a calender via http
Hi,

now I'm totally surprised:
I already fixed the contact stuff (together with thangus) and now it's the same for calendar.

It is not possible to create a calendar from 'outside' via html.

Since I know now where I have to look, I found out very fast, that the function

createCalendar() in lib/sabre/backend.php

is a total 'disaster'.
All parameters are put into $values and later $newValues is used, which is never defined before
Next point:
All keys were done with a ':' in front, but later the access is without ':'

To make it short:

Here is a working function:

        public function createCalendar($principalUri,$calendarUri, array $properties) {
                $values = array(
                        'principaluri' => $principalUri,
                        'uri'             => $calendarUri,
                        'ctag'           => 1,
                );

                // Default value
                $sccs = '{urn:ietf:params:xml:ns:caldav}supported-calendar-component-set';
                if (!isset($properties[$sccs])) {
                        $values['components'] = 'VEVENT,VTODO';
                } else {
                        if (!($properties[$sccs] instanceof Sabre_CalDAV_Property_SupportedCalendarComponentSet)) {
                                throw new Sabre_DAV_Exception('The ' . $sccs . ' property must be of type: Sabre_CalDAV_Property_SupportedCalendarComponentSet');
                        }
                        $values['components'] = implode(',',$properties[$sccs]->getValue());
                }

                foreach($this->propertyMap as $xmlName=>$dbName) {
                        if (isset($properties[$xmlName])) {
                                $values[$dbName] = $properties[$xmlName];
                        }
                }

                if(!isset($values['displayname'])) $values['displayname'] = 'unnamed';
                if(!isset($values['components'])) $values['components'] = 'VEVENT,VTODO';
                if(!isset($values['timezone'])) $values['timezone'] = null;
                if(!isset($values['calendarorder'])) $values['calendarorder'] = 0;
                if(!isset($values['calendarcolor'])) $values['calendarcolor'] = null;
                if(!is_null($values['calendarcolor']) && strlen($values['calendarcolor']) == 9) {
                        $values['calendarcolor'] = substr($values['calendarcolor'], 0, 7);
                }

                return OC_Calendar_Calendar::addCalendarFromDAVData($principalUri,$calendarUri,$values['displayname'],$values['components'],$values['timezone'],$values['calendarorder'],$values['calendarcolor']);
        }

And I think $fieldname can totally removed and  $myValue too
They makes no sense, but I'm not familar with php, please check this.
I checked it with my code: it works without them, so I removed the lines.

Best regards

Bernd

And greetings to thangus!

### PR DESCRIPTION
**Pull request by ednt**
Fixes #350

Not possible to create a calendar via http.

### CODE REVIEW

**Comment by tanghus** [comment]
@ednt What tool do you use to create calendars? I use KDE/Akonadi and can't create collections on a CalDAV server - I need to file a bug report on that ;)

**Comment by ednt** [comment]
Hi tanghus,

hey, also calendar? I thought your are the contact man :)

As usual:
calendartool
calendartool_x64
calendartool exe

I'm also surprised, that so less tools and programs are available for Cal- and CardDAV.
That's the reason why I have to write my own stuff.

Bernd

**Comment by wtayyeb** [comment]
> that so less tools and programs are available for Cal- and CardDAV.

[eM Client](http://emclient.com/) is Free desktop app, and completely support Cal-DAV and CardDAV
Thunderbird+Lightning + some tricks will support Cal-DAV and CardDAV

**Comment by georgehrke** [comment]
free as in free beer, but not free as in free speech hehe

**Comment by ednt** [comment]
I don't need an other e-mail program.
Lightning does not work 100% right.
It looks like it is not reliable if you updated your owncloud calendar entry.
Maybe next day it's right, but not during 1 hour.

I didn't found a 'standalone' calendar with CalDAV or a 'standalone' addressbook with CardDAV.

**Comment by tanghus** [comment]
> hey, also calendar? I thought your are the contact man :)

@ednt Just filling in while @georgehrke is reading for exams :)

**Comment by tanghus** [comment]
@ednt Running `./calendartool http://localhost/owncloud/remote.php/caldav/calendars/tanghus/ -l -v -u user -l pwd` I get `Sorry, only http:// possible.`.
Reordering the options doesn't help.

**Comment by tanghus** [comment]
I just tried with T-Bird/Lightning, and it seems it doesn't even try to add the calendar?

**Comment by ednt** [comment]
Hi,

./calendartool works if you use -p in front of the password :)

If you have no -x in front of a parameter it thinks that thats the url.

Btw. you should only use
./calendartool http://localhost/owncloud -u user -p password -l

the full path for the user is build by the program.
And remote.php ... is also fix inside the code.
You can use this tools only for owncloud.

**Comment by ednt** [comment]
Hi,

forgot to implement the verbose messages :)
with -v -v you get now the content of the request and the answer.

calendartool
calendartool_x64
calendartool exe

**Comment by tanghus** [comment]
@ednt I just tested with your calendartool on git master - should be the same as stable6 - and calendar creation works fine. Changing the display name doesn't but that's another matter - or maybe you have time to look at that too?
For pure code cleaning your patch is still valid though. I'll add a few hints in that department now you're there.

File: lib/sabre/backend.php [hunk 1]
```diff
@@ -92,48 +92,41 @@ public function getCalendarsForUser($principalUri) {
...
 			if (isset($properties[$xmlName])) {
-
-				$myValue = $properties[$xmlName];
-				$values[':' . $dbName] = $properties[$xmlName];
-				$fieldNames[] = $dbName;
+				$values[$dbName] = $properties[$xmlName];
 			}
 		}
 
-		if(!isset($newValues['displayname'])) $newValues['displayname'] = 'unnamed';
```
Comment (tanghus) [inline lib/sabre/backend.php:127]:
``` php
if(!isset($newValues['displayname'])) {
    $newValues['displayname'] = 'unnamed';
}
```

Our coding guidelines says to always use curlies. This is old code so there are still a lot of them missing as you can see.

File: lib/sabre/backend.php [hunk 2]
```diff
@@ -92,48 +92,41 @@ public function getCalendarsForUser($principalUri) {
...
+		if(!isset($values['components'])) $values['components'] = 'VEVENT,VTODO';
+		if(!isset($values['timezone'])) $values['timezone'] = null;
+		if(!isset($values['calendarorder'])) $values['calendarorder'] = 0;
+		if(!isset($values['calendarcolor'])) $values['calendarcolor'] = null;
+		if(!is_null($values['calendarcolor']) && strlen($values['calendarcolor']) == 9) {
+			$values['calendarcolor'] = substr($values['calendarcolor'], 0, 7);
 		}
 
-		return OC_Calendar_Calendar::addCalendarFromDAVData($principalUri,$calendarUri,$newValues['displayname'],$newValues['components'],$newValues['timezone'],$newValues['calendarorder'],$newValues['calendarcolor']);
+		return OC_Calendar_Calendar::addCalendarFromDAVData($principalUri,$calendarUri,$values['displayname'],$values['components'],$values['timezone'],$values['calendarorder'],$values['calendarcolor']);
```
Comment (tanghus) [inline lib/sabre/backend.php]:
``` php
return OC_Calendar_Calendar::addCalendarFromDAVData(
    $principalUri,
    $calendarUri,
    $values['displayname'],
    $values['components'],
    $values['timezone'],
    $values['calendarorder'],
    $values['calendarcolor']
);
```

Readability and line length.

File: lib/sabre/backend.php [hunk 3]
```diff
@@ -92,48 +92,41 @@ public function getCalendarsForUser($principalUri) {
...
-		if(!isset($newValues['calendarorder'])) $newValues['calendarorder'] = 0;
-		if(!isset($newValues['calendarcolor'])) $newValues['calendarcolor'] = null;
-		if(!is_null($newValues['calendarcolor']) && strlen($newValues['calendarcolor']) == 9) {
-			$newValues['calendarcolor'] = substr($newValues['calendarcolor'], 0, 7);
+		if(!isset($values['displayname'])) $values['displayname'] = 'unnamed';
+		if(!isset($values['components'])) $values['components'] = 'VEVENT,VTODO';
+		if(!isset($values['timezone'])) $values['timezone'] = null;
+		if(!isset($values['calendarorder'])) $values['calendarorder'] = 0;
+		if(!isset($values['calendarcolor'])) $values['calendarcolor'] = null;
+		if(!is_null($values['calendarcolor']) && strlen($values['calendarcolor']) == 9) {
```
Comment (tanghus) [inline lib/sabre/backend.php]: Always use `===` for comparisons.

**Comment by ednt** [comment]
Hi,

if I add a new calendar against 6.0.2, the calendar is created, but the displayname is not set.
It is 'unnamed'.
But:

``` XML
<?xml version="1.0" encoding="utf-8" ?>
<C:mkcalendar xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
 <D:set>
  <D:prop>
   <D:displayname>test4</D:displayname>
   <C:calendar-description>Description of test4</C:calendar-description>
   <C:supported-calendar-component-set>
    <C:comp name="VEVENT"/>
    <C:comp name="VTODO"/>
    <C:comp name="VJOURNAL"/>
   </C:supported-calendar-component-set>
  </D:prop>
 </D:set>
</C:mkcalendar>
```

Definately sets the displayname.

Also look at
https://github.com/owncloud/calendar/issues/350

**Comment by ednt** [comment]
Not updateing the displayname comes from a missing description entry:

``` PHP
public $propertyMap = array(
        '{DAV:}displayname' => 'displayname',
        '{urn:ietf:params:xml:ns:caldav}calendar-timezone' => 'timezone',
        '{urn:ietf:params:xml:ns:caldav}calendar-description' => 'description',
        '{http://apple.com/ns/ical/}calendar-order' => 'calendarorder',
        '{http://apple.com/ns/ical/}calendar-color' => 'calendarcolor',
    ); 
```

And to avoid a logfile entry:
add $row['description'] = ''; in lib/calendar.php

``` PHP
while( $row = $result->fetchRow()) {
            $row['permissions'] = OCP\PERMISSION_CREATE
                | OCP\PERMISSION_READ | OCP\PERMISSION_UPDATE
                | OCP\PERMISSION_DELETE | OCP\PERMISSION_SHARE;
            $row['description'] = '';
            $calendars[] = $row;
        }
```

But I have wriiten this already here ... somewhere.

**Comment by georgehrke** [comment]
@tanghus Is this pr still necessary after https://github.com/owncloud/calendar/commit/848795deb8094c3cd05e8559c50e2505434c38bc ?

**Comment by tanghus** [comment]
@georgehrke not really sure tbh.

**Comment by ednt** [comment]
The changes are not in the master branch,
and it is still not possible without manual patching (6.0.3).

File: lib/sabre/backend.php (hunk 1 above)
Comment (georgehrke) [inline lib/sabre/backend.php:127]: @ednt please fix :) (this and the other issues tanghus mentioned as well plz)

**Comment by libasys** [comment]
Tested the changes and it works. Good job @ednt. Thanks!

**Comment by ednt** [comment]
I'm deeply disappointed.

6.0.4 is out and half of the bugs which I fixed since 6.0.1 are still present.
I can not do more than I had already done.

It looks like that I have to fix this bugs for each new version for the rest of the time we use ownCloud.
Not what I really want to do.
This means 3 to 4 hours work after each new version.

**Comment by georgehrke** [comment]
@ednt @tanghus noted a various issues that haven't been fixed yet in this pr

**Comment by georgehrke** [comment]
@ednt I'll merge once the issues we mentioned above are fixed :)

**Comment by ednt** [comment]
I just read backwards, but I didn't found 'mentioned issues'.
If I have to do something, please tell me what and I'll do it.
Or do you mean other fixes which don't belong to this stuff?

**Comment by georgehrke** [comment]
see tanghus's comments here: https://github.com/owncloud/calendar/pull/354/files

**Comment by ednt** [comment]
That's not nice.

The mentioned issues were not from me, they were already as they are.
So I have to fix stuff from 'programming guidelines' I never heard of.
And also I have nothing todo with Git (I told this already tanghus).
So I'm never sure if the stuff I did is ok or not.

And in the time tanghus has written the comments, he had also changed the stuff in the same time.
Why he didn't done this?

But I changed now something 'as requested'.
I hope it's right, it was in a patch-2 branch, the first branch I found with my changes.

I think in the future I will only report bugs and bump them if not fixed.
That's easier for me.

**Comment by ednt** [comment]
Hi,

first good news: 7.0 works as expected.

But....

If you create a calendar all optional values have the default values.
You use the array $newValues, but from where is it filled ???

backend.php line 129 to 136

Bernd

**Comment by libasys** [comment]
Where are the $newValues variable is used with this patch? I see no variable $newValues! The new array is called $values!

**Comment by ednt** [comment]
Hi,

I downloaded 
https://download.owncloud.org/community/owncloud-7.0.0.tar.bz2

In
\owncloud-7.0.0.tar\owncloud\apps\calendar\lib\sabre\backend.php

at line 129 to 136 you can find $newValues which is never used before.

Bernd

**Comment by libasys** [comment]
ah ok now i understand this pr is not merged and the old is in use! the $newValues array created automatically. The correct way is to add on the top of the method and define: $newValues=array(); so this is better to understand!

**Comment by DeepDiver1975** [comment]
closing this in favor of #953

**Pull request closed without merging**